src/Core/Services/Implementations/ProviderService.cs

```diff
@@ -0,0 +1,346 @@
+                }
+                catch (BadRequestException e)
+                {
+                    result.Add(Tuple.Create(providerUser, e.Message));
+                }
+            }
+
+            await _eventService.LogProviderUsersEventAsync(events);
+
+            return result;
+        }
+
+        public async Task SaveUserAsync(ProviderUser user, Guid savingUserId)
+        {
+            if (user.Id.Equals(default))
+            {
+                throw new BadRequestException("Invite the user first.");
+            }
+
+            if (user.Type != ProviderUserType.ProviderAdmin &&
+                !await HasConfirmedProviderAdminExceptAsync(user.ProviderId, new[] {user.Id}))
+            {
+                throw new BadRequestException("Provider must have at least one confirmed ProviderAdmin.");
+            }
+
+            await _providerUserRepository.ReplaceAsync(user);
+            await _eventService.LogProviderUserEventAsync(user, EventType.ProviderUser_Updated);
+        }
+
+        public async Task<List<Tuple<ProviderUser, string>>> DeleteUsersAsync(Guid providerId,
+            IEnumerable<Guid> providerUserIds, Guid deletingUserId)
+        {
+            var providerUsers = await _providerUserRepository.GetManyAsync(providerUserIds);
+
+            if (!await HasConfirmedProviderAdminExceptAsync(providerId, providerUserIds))
+            {
+                throw new BadRequestException("Provider must have at least one confirmed ProviderAdmin.");
+            }
+
+            var result = new List<Tuple<ProviderUser, string>>();
+            var deletedUserIds = new List<Guid>();
+            var events = new List<(ProviderUser, EventType, DateTime?)>();
+
+            foreach (var providerUser in providerUsers)
+            {
+                try
+                {
+                    if (providerUser.ProviderId != providerId)
+                    {
+                        throw new BadRequestException("Invalid user.");
+                    }
+                    if (providerUser.UserId == deletingUserId)
+                    {
+                        throw new BadRequestException("You cannot remove yourself.");
+                    }
+
+                    events.Add((providerUser, EventType.ProviderUser_Removed, null));
+
+                    result.Add(Tuple.Create(providerUser, ""));
+                    deletedUserIds.Add(providerUser.Id);
+                }
+                catch (BadRequestException e)
+                {
+                    result.Add(Tuple.Create(providerUser, e.Message));
+                }
+
+                await _providerUserRepository.DeleteManyAsync(deletedUserIds);
+            }
+
+            await _eventService.LogProviderUsersEventAsync(events);
+
+            return result;
+        }
+
+        public Task AddOrganization(Guid providerId, Guid organizationId, Guid addingUserId, string key) => throw new NotImplementedException();
```

Comment:
Please add `//TODO` comments in here if we're intending to implement these later on.